### PR TITLE
Context

### PR DESCRIPTION
**Pull request by junchaowu**
Add stack context into tchannel. 
Removed the proxy class.
Make zipkintracehook not default.

### CODE REVIEW

File: UPGRADE.rst [hunk 1]
```diff
@@ -31,6 +31,7 @@ From 0.15 to 0.16
       def endpoint(request):
           return Response({'resp': 'body'})
 
+- Stop using `` from tchannel.tornado import TChannel``
```
Comment (abhinav) [inline UPGRADE.rst]: what

File: tchannel/context.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from __future__ import absolute_import
+import threading
+
+
+class TChannelLocal(threading.local):
+    __slots__ = ('context',)
+
```
Comment (abhinav) [inline tchannel/context.py]: Why is this a class if it's not doing anything?

You can just do,

    _LOCAL = threading.local()
    _LOCAL.context = None

File: tchannel/context.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+class Context(object):
+    """Stack context for each TChannel instance."""
+
+    __slots__ = ('parent_tracing', '_old_context',)
+
+    def __init__(self):
+        self.parent_tracing = None
+
+    def __enter__(self):
+        self._old_context = _LOCAL.context
```
Comment (abhinav) [inline tchannel/context.py]: If you `__enter__` in a different thread, this will fail because `_LOCAL.context` will not exist. The `context` attribute exists only in the thread that set it. You have to check with `hasattr` first.

File: tchannel/context.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    def __init__(self):
+        self.context = None
+
+_LOCAL = TChannelLocal()
+_LOCAL.context = None
+
+
+class Context(object):
+    """Stack context for each TChannel instance."""
```
Comment (abhinav) [inline tchannel/context.py]: Pick a more descriptive name.

Also, this needs a more detailed docstring. The docstring should answer what the purpose of this class is.

File: tchannel/context.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    __slots__ = ('parent_tracing', '_old_context',)
+
+    def __init__(self):
+        self.parent_tracing = None
```
Comment (abhinav) [inline tchannel/context.py]: Initialize all attributes in `__init__`, including `_old_context`.

File: tchannel/context.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    def __enter__(self):
+        self._old_context = _LOCAL.context
+        _LOCAL.context = self
+
+    def __exit__(self, type, value, traceback):
+        _LOCAL.context = self._old_context
+
+
+def get_local():
+    return _LOCAL
```
Comment (abhinav) [inline tchannel/context.py]: Don't expose the actual thread local outside this module.

Comment (abhinav) [inline tchannel/context.py]: Instead, expose a method that returns the 'current context'

File: tchannel/context.py (hunk 4 above)
Comment (abhinav) [inline tchannel/context.py]: Also this isn't really a stack context. I see no mention of stack context anywhere in this module.

File: tchannel/context.py (hunk 2 above)
Comment (junchaowu) [inline tchannel/context.py]: so default the context will be part of the thread local.

File: tchannel/context.py (hunk 3 above)
Comment (junchaowu) [inline tchannel/context.py]: context will exist. It is part of the thread local class.

File: tchannel/tchannel.py [hunk 7]
```diff
@@ -57,6 +60,15 @@ def __init__(self, name, hostport=None, process_name=None,
...
         self.thrift = schemes.ThriftArgScheme(self)
 
+        self._local = get_local()
+
+    def get_context(self):
+        """
+
+        :return: stack context in current running aysnc thread.
+        """
+        return self._local.context
```
Comment (abhinav) [inline tchannel/tchannel.py]: No. Don't access the thread local directly.

File: tchannel/tornado/dispatch.py [hunk 8]
```diff
@@ -182,9 +185,11 @@ def handle_call(self, request, connection):
                 )
 
                 # get new top-level resp from controller
-                new_resp = yield gen.maybe_future(
-                    handler.endpoint(new_req)
-                )
+                with StackContext(lambda: _context):
+                    tchannel.get_context().parent_tracing = request.tracing
+                    f = handler.endpoint(new_req)
+
```
Comment (abhinav) [inline tchannel/tornado/dispatch.py]: This should all be encapsulated in that same module that manages the thread. Ideally, this code block should just be

    with Context(parent_tracing):
        # ...

(With a better name than Context, of course.)

File: tchannel/tornado/dispatch.py [hunk 9]
```diff
@@ -287,37 +282,3 @@ def not_found(request, response, proxy):
                 request.service,
             ),
         )
-
-
-class TChannelProxy(object):
-    """TChannel Proxy with additional runtime info
```
Comment (abhinav) [inline tchannel/tornado/dispatch.py]: YASS

File: tchannel/context.py (hunk 6 above)
Comment (junchaowu) [inline tchannel/context.py]: TChannel will call this method to get the local. other places will call tchannel.get_context to get context.

File: tchannel/tornado/dispatch.py (hunk 8 above)
Comment (junchaowu) [inline tchannel/tornado/dispatch.py]: Parent_tracing will be part of context. context != parent_tracing.

File: tchannel/tornado/peer.py [hunk 10]
```diff
@@ -453,6 +453,9 @@ def __init__(self,
         self.tchannel = peer_group.tchannel
         self.service = service
         self.parent_tracing = parent_tracing
+        if self.tchannel.get_context():
+            self.parent_tracing = (parent_tracing or
+                                   self.tchannel.get_context().parent_tracing)
```
Comment (abhinav) [inline tchannel/tornado/peer.py]: nit: could be cleaner

    if not self.parent_tracing and self.tchannel.get_context():
        self.parent_tracing = self.tchannel.get_context().parent_tracing

Also, I'm not sure if `get_tracing` needs to exist. what's the use case for people needing to access tho context object?

File: tchannel/tornado/tchannel.py [hunk 11]
```diff
@@ -121,9 +122,6 @@ def __init__(self, name, hostport=None, process_name=None,
         self.event_emitter = EventEmitter()
         self.hooks = EventRegistrar(self.event_emitter)
 
-        from ..zipkin.zipkin_trace import ZipkinTraceHook
-        self.hooks.register(ZipkinTraceHook(tchannel=self))
```
Comment (abhinav) [inline tchannel/tornado/tchannel.py]: Cool so we can drop the hard dependency on thrift?

File: tests/integration/thrift/test_tornado_client.py [hunk 12]
```diff
@@ -82,7 +83,7 @@ def test_thrift_exception(mock_server, thrift_service):
         'thrift',
         method='getItem',
     ).and_raise(thrift_service.ItemDoesNotExist("stahp"))
-
+    mock_server.tchannel.hooks.register(ZipkinTraceHook(tchannel=mock_server))
```
Comment (abhinav) [inline tests/integration/thrift/test_tornado_client.py]: If this test isn't testing zipkin integration, why does it need the hook?

Comment (junchaowu) [inline tests/integration/thrift/test_tornado_client.py]: The existing test has this mock_trac_record.called assertion. I guess it is testing if trace will be triggered when exception happens in thrift. This is your test. If it doesn't mean that. I will delete it.

File: tchannel/context.py (hunk 2 above)
Comment (abhinav) [inline tchannel/context.py]: Ah, if you create it like this it'll never be missing the attribute, even in other threads. Okay.

Comment (abhinav) [inline tchannel/context.py]: Deleting comment.

File: tchannel/context.py (hunk 6 above)
Comment (abhinav) [inline tchannel/context.py]: The thread local and the stack context should not be exposed outside this module. Thread local provides one level of 'local'-ness and the stack context provides the other. These are both details that other parts of the system don't care about. They should be able to do `get_current_context` and get the 'current' `Context` object.

Comment (abhinav) [inline tchannel/context.py]: And I don't see why `TChannel.get_context` exists either. Users don't need to be able to access the context object. We're managing it transparently for them.

File: tchannel/tornado/dispatch.py (hunk 8 above)
Comment (abhinav) [inline tchannel/tornado/dispatch.py]: Yes, I know. `parent_tracing` is tracked by context. It is part of the state that it tracks per 'async thread'. I'm assuming `Context` is the class that will hold all this state. Right now, you'll do 

    with Context(parent_tracing):
      # ...

But if we decide to add another piece of state to be tracked, you'd basically change it to

    with Context(parent_tracing, second_state):
        # ...

File: tests/integration/thrift/test_tornado_client.py (hunk 12 above)
Comment (abhinav) [inline tests/integration/thrift/test_tornado_client.py]: Existing tests may have been asserting zipkin behavior because we had it enabled by default previously.

Now that it's opt-in, it might make sense to eliminate these references.
Although, we would still want an integration test of some kind

File: tchannel/tornado/dispatch.py (hunk 9 above)
Comment (breerly) [inline tchannel/tornado/dispatch.py]: :laughing: :cocktail:

File: tchannel/tornado/tchannel.py (hunk 11 above)
Comment (breerly) [inline tchannel/tornado/tchannel.py]: Let's do that in another PR

File: CHANGES.rst [hunk 13]
```diff
@@ -19,6 +19,8 @@ Changelog
   the new server API. See the upgrade guide for details.
 - **BREAKING** - remove ``retry_delay`` option in the ``tchannel.tornado.send``
   method.
+- **BREAKING** - remove third arugment ``proxy`` from the server handler interface.
+- **BREAKING** - make the zipkintracehook not default registered.
```
Comment (blampe) [inline CHANGES.rst]: ``ZipkinTraceHook`` is not longer registered by default.

File: examples/benchmark/thrift/server.py [hunk 14]
```diff
@@ -34,7 +34,7 @@
 
 
 @app.register(KeyValue)
-def getValue(request, response, tchannel):
+def getValue(request, response):
```
Comment (breerly) [inline examples/benchmark/thrift/server.py]: If we're going to update this example let's switch over to `tchannel.TChannel` instead of `tchannel.tornado.TChannel`.

File: examples/benchmark/thrift/server.py [hunk 15]
```diff
@@ -45,7 +45,7 @@ def getValue(request, response, tchannel):
 
 
 @app.register(KeyValue)
```
Comment (breerly) [inline examples/benchmark/thrift/server.py]: Let's call this `tchannel` instead of `app`.

File: examples/benchmark/thrift/server.py (hunk 14 above)
Comment (blampe) [inline examples/benchmark/thrift/server.py]: +1

File: tchannel/context.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+class TChannelLocal(threading.local):
+    __slots__ = ('context',)
+
+    def __init__(self):
+        self.context = None
+
+_LOCAL = TChannelLocal()
+
+
+class TChannelContext(object):
```
Comment (breerly) [inline tchannel/context.py]: Let's rename to something more obvious... I think `RequestContext` maybe...

File: tchannel/testing/vcr/server.py [hunk 17]
```diff
@@ -84,7 +84,7 @@ def __init__(self, cassette, unpatch):
         VCRProxy.VCRServiceError,
     ))
     @gen.coroutine
-    def send(self, request, response, channel):
+    def send(self, request, response):
```
Comment (breerly) [inline tchannel/testing/vcr/server.py]: hmmm... so I guess we'll have to get the sync client and vcr working with the top-level `tchannel.TChannel` too

File: tests/integration/thrift/test_tornado_client.py [hunk 18]
```diff
@@ -66,13 +65,7 @@ def test_protocol_error(mock_server, thrift_service):
 
     client = mk_client(thrift_service, mock_server.port, trace=False)
     with pytest.raises(errors.ProtocolError):
-        with patch(
-            'tchannel.zipkin.tracers.TChannelZipkinTracer.record',
```
Comment (blampe) [inline tests/integration/thrift/test_tornado_client.py]: Looks like we lost test coverage on the zipkin hooks, can we move these to test cases that intentionally opt-in to zipkin?

File: tests/integration/trace/test_zipkin_trace.py [hunk 19]
```diff
@@ -127,7 +124,7 @@ def test_zipkin_trace(trace_server):
 
 
 @pytest.mark.gen_test
-def test_tcollector_submit(monkeypatch, trace_server):
+def test_tcollector_submit(trace_server):
```
Comment (blampe) [inline tests/integration/trace/test_zipkin_trace.py]: Don't we need to register the zipkin event hooks?

**Comment by breerly** [comment]
Didn't we want to attach `context` to `tchannel.Request`?

```
@tchannel.json.register
def endpoint(request):
    print request.context    # tchannel.context.RequestContext
```

File: tests/integration/thrift/test_tornado_client.py (hunk 18 above)
Comment (junchaowu) [inline tests/integration/thrift/test_tornado_client.py]: No. We have test_zipkin_trace.py to test zipkin hooks.

File: tests/integration/trace/test_zipkin_trace.py (hunk 19 above)
Comment (junchaowu) [inline tests/integration/trace/test_zipkin_trace.py]: No. This unit test is test zipkin trace's submit method. It is not an end to end test. We have 'test_zipkin_trace' test doing end to end test for zipkin. In this test, we register zipkintracehook.

**Comment by junchaowu** [comment]
@breerly whether we want to put context in the request object or whether we take context as argument, need further discussion. At this init stage, I would keep scope of this PR. 

Besides if we decide update all the example or test to use new tchannel.TChannel, that's a large amount code change. Let's put it in a separate PR.

File: tests/test_context.py [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+import pytest
+from tornado import gen
+
+from tchannel import TChannel
+from tchannel import Response
+from tchannel.context import get_current_context
+
+
+@pytest.mark.gen_test
+def test_context():
```
Comment (breerly) [inline tests/test_context.py]: `s/test_context/test_context_should_carry_tracing_info/`

Comment (breerly) [inline tests/test_context.py]: This test should be in `test_tchannel.py` since you're doing an integration test through the public API. This file I would expect to test just the pieces in `context.py`.

File: tchannel/tornado/dispatch.py [hunk 21]
```diff
@@ -182,9 +184,10 @@ def handle_call(self, request, connection):
                 )
 
                 # get new top-level resp from controller
-                new_resp = yield gen.maybe_future(
-                    handler.endpoint(new_req)
-                )
+                with StackContext(lambda: RequestContext(request.tracing)):
+                    f = handler.endpoint(new_req)
```
Comment (abhinav) [inline tchannel/tornado/dispatch.py]: Can we also move the StackContext management to the context module? As said previously, I would like for those details to not be leaked outside that module. This block of code should ideally look like this:

    with RequestContext(request.tracing):
        # ...

Or something similar.

File: tchannel/context.py [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+        _LOCAL.context = self
+
+    def __exit__(self, type, value, traceback):
+        _LOCAL.context = self._old_context
+
+
+def get_current_context():
+    """
+
+    :return: stack context in current running aysnc thread.
```
Comment (abhinav) [inline tchannel/context.py]: You're not returning a stack context. You're returning a request context.

File: tests/test_context.py [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+        scheme=schemes.RAW,
+        service='server',
+        arg1='endpoint1',
+        arg2='req headers',
+        arg3='req body',
+        hostport=server.hostport,
+    )
+
+    assert context[0].parent_tracing.name == 'endpoint1'
+    assert context[1].parent_tracing.name == 'endpoint2'
```
Comment (abhinav) [inline tests/test_context.py]: Also match the trace ID (since it's part of the same request)

**Comment by abhinav** [comment]
Couple minor issues but the core functionality looks good.

**Comment by abhinav** [comment]
looks good besides comment nit.
